### PR TITLE
Refactor metrics

### PR DESCRIPTION
**Pull request by mumichae**
Clean up LISI and kBET metrics.

- split iLISI and cLISI
- kBET scaling inside function
- Move kBET NeighborsError to kBET function

### CODE REVIEW

**Comment by LuckyMD** [comment]
Did we really want to split cLISI and iLISI? These are computed in the same function, no?

**Comment by mbuttner** [comment]
> Did we really want to split cLISI and iLISI? These are computed in the same function, no?

I wouldn't split cLISI and iLISI, either. In fact, I would rather merge the computation of both cLISI and iLISI, because the steps are the same, just a different label to compute the metric on - it would save quite some computation time.

**Comment by mumichae** [comment]
From what I can see, the only time saved is during recomputing the kNN graph. If we move the control of preprocessing outside of the `ilisi_graph` and `clisi_graph` functions, that time could be halved. Also, I didn't think that was the most time consuming step (but would have to double-check). The actual LISI computation (using the C code) is done separately for each metric anyway, so splitting up the code shouldn't change much.

Additionally, the old `lisi_graph` function works as before, computing both scores, so we could use that for the `metrics` function if we wanted. I just thought it would be best to have the same level of control over `iLISI` an `cLISI` as we would have over all the other metrics when using the API.

**Comment by mbuttner** [comment]
> From what I can see, the only time saved is during recomputing the kNN graph. If we move the control of preprocessing outside of the `ilisi_graph` and `clisi_graph` functions, that time could be halved. Also, I didn't think that was the most time consuming step (but would have to double-check). In the original implementation all the LISI computations (using the C code) are called separately for `iLISI` and `cLISI` anyway, so splitting up the code shouldn't change much.
> 
> Additionally, the old `lisi_graph` function works as before, computing both scores, so we could use that for the `metrics` function if we wanted. I just thought it would be best to have the same level of control over `iLISI` an `cLISI` as we would have over all the other metrics when using the API.

The thing with the two LISI functions is that the `batch_key`/`label_key` info is only used at the very end of `compute_simpson_index_graph`:
```
batch = batch_labels[knn_idx]
B = convertToOneHot(batch, n_batches)
sumP = np.matmul(P, B)  # sum P per batch
simpson[i[0]] = np.dot(sumP, sumP)  # sum squares
```
So the most efficient way to compute both `iLISI` and `cLISI` would be to add both batch and label keys at this piece of code. However, then both functions are invariably intertwined.

File: scIB/metrics/lisi.py
Comment (mbuttner) [reply scIB/metrics/lisi.py:248]: Ah OK, indeed!

File: scIB/metrics/kbet.py
Comment (mumichae) [reply scIB/metrics/kbet.py:161]: Thanks for noticing

File: tests/metrics/test_ilisi.py
Comment (mumichae) [reply tests/metrics/test_ilisi.py:4]: Yes, the CI runs through without error

**Comment by mumichae** [comment]
> The thing with the two LISI functions is that the `batch_key`/`label_key` info is only used at the very end of `compute_simpson_index_graph`:
> 
> ```
> batch = batch_labels[knn_idx]
> B = convertToOneHot(batch, n_batches)
> sumP = np.matmul(P, B)  # sum P per batch
> simpson[i[0]] = np.dot(sumP, sumP)  # sum squares
> ```
> 
> So the most efficient way to compute both `iLISI` and `cLISI` would be to add both batch and label keys at this piece of code. However, then both functions are invariably intertwined.

So that means that the old implementation of the score wasn't the most efficient either, because we call the `lisi_graph_py` and with that the `compute_simpson_index_graph` twice anyway. 

Would it be better if we extracted the `compute_simpson_index_graph` functionality outside of `lisi_graph_py`, compute it once and then pass the output for cLISI and iLISI computations?
In that case I'd suggest we optimise the `lisi_graph` function, but still keep separate cLISI and iLISI functions for the API. The latter could have an option to preprocess the simpson index and pass it to those functions. I hope that isn't overkill...

**Comment by mumichae** [comment]
I think this PR should be ready to merge and we can address LISI computation optimisations in a different PR. Any comments or wishes to the current state?

**Comment by gokceneraslan** [comment]
This PR also fixes, #266, so it'd be really great to see it merged 😄

**Comment by mumichae** [comment]
@mbuttner Could you go over the latest changes and approve if the PR is ready to be merged?

File: scIB/metrics/metrics.py
Comment (LuckyMD) [inline scIB/metrics/metrics.py:160]: Could add a deprecation comment for this as well... we don't use it anymore as it only works for python methods.

File: scIB/metrics/metrics.py
Comment (LuckyMD) [inline scIB/metrics/metrics.py:187]: I thought this parameter was removed from the kBET method?

File: scIB/metrics/metrics.py
Comment (LuckyMD) [inline scIB/metrics/metrics.py:216]: why remove verbosity passing here?

File: scIB/metrics/metrics.py
Comment (LuckyMD) [inline scIB/metrics/metrics.py:351]: any reason to not just keep this in `lisi_graph_`? I guess this also allows separate clisi and ilisi runs?

**Review by LuckyMD — COMMENTED** [review]
I love the lisi tests! Just 1 or 2 questions... otherwise this looks good from my side. If you answer the questions, we can merge.

File: scIB/metrics/metrics.py
Comment (LuckyMD) [reply scIB/metrics/metrics.py:187]: this is only passed to LISI now?

File: scIB/metrics/metrics.py
Comment (mumichae) [reply scIB/metrics/metrics.py:160]: Deprecation for the LISI parameters?

File: scIB/metrics/metrics.py
Comment (mumichae) [reply scIB/metrics/metrics.py:187]: yes, that's why I renamed it. We never used it for kBET.

File: scIB/metrics/metrics.py
Comment (mumichae) [reply scIB/metrics/metrics.py:216]: The output wasn't very informative (it's only preprocessing), so I removed it

File: scIB/metrics/metrics.py
Comment (mumichae) [reply scIB/metrics/metrics.py:351]: Yes, the user then also has the option to run one and not the other. I just added that for consistency with the rest of the behaviour. The old `lisi_graph_` parameter should just work as before

**Review by LuckyMD — APPROVED** [review]

File: scIB/metrics/metrics.py
Comment (LuckyMD) [reply scIB/metrics/metrics.py:197]: Maybe it's sth for a future "debug" option for verbosity. Let's leave for now.

**Comment by LuckyMD** [comment]
Once @mbuttner reviews, you can merge. All good from my side now.

**Review by mbuttner — APPROVED** [review]
Looks good to me.

**Pull request merged**
